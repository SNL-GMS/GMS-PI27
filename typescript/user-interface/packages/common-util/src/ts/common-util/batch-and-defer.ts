import debounce from 'lodash/debounce';
import defer from 'lodash/defer';

/**
 * Will return a version of the input function that will batch all given parameters into
 * fewer calls.
 *
 * @param fn the function to batch and defer
 * @param wait how long to wait for more params (default 150)
 * @returns the batched function
 */
export const batchAndDefer = <T>(fn, wait = 150) => {
  const batchedArguments: unknown[] = [];

  const runFn = () => {
    // Clone of the batched arguments so defer will still have access within this scope
    const args = [...batchedArguments];
    // empty batched arguments
    batchedArguments.length = 0;

    defer(fn, args);
  };

  const debounced = debounce(runFn, wait);

  return (incomingArgs: T) => {
    if (Array.isArray(incomingArgs)) {
      batchedArguments.push(...incomingArgs);
    } else {
      batchedArguments.push(incomingArgs);
    }

    debounced.cancel();
    debounced();
  };
};
