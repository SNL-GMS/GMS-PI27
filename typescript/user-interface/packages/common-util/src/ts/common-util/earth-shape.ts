import { toRadians } from './unit-util';

/**
 * This is a TypeScript version of the EarthShape Java class
 *
 * @see gms-common/java/gms/shared/feature-prediction/feature-prediction-utilities/src/main/java/gms/shared/featureprediction/utilities/math/EarthShape.java
 */

const EARTH_SPHERE_RADIUS = 6371;
const EARTH_ELLIPSOID_RADIUS = 6378.137;
const EARTH_ELLIPSOID_IERS_RADIUS = 6378.1366;
const IERS_INVERSE_FLATTENING_FIVE = 298.25642;
const GRS80_INVERSE_FLATTENING = 298.257222101;
const WGS84_INVERSE_FLATTENING = 298.257223563;

class EarthShape {
  /**
   * True for EarthShapes that assume that the Earth has constant radius for purposes of converting
   * between radius and depth.
   */
  public constantRadius: boolean;

  /** The radius of the earth at the equator. */
  public equatorialRadius: number;

  /** flattening equals [ 1 - b/a ] where a is equatorial radius and b is the polar radius */
  public flattening: number;

  /** [ 1./flattening ] */
  public inverseFlattening: number;

  /**
   * Eccentricity squared.
   *
   * <p>Equals [ flattening * (2. - flattening) ]
   *
   * <p>Also equals [ 1 - sqr(b)/sqr(a) ] where a is the equatorial radius and b is the polar
   * radius.
   */
  public eccentricitySqr: number;

  /** Equals [ 1 - eccentricitySqr ] */
  public e1: number;

  /** Equals [ eccentricitySqr / (1 - eccentricitySqr) ] */
  public e2: number;

  /**
   * constructor
   *
   * @param inverseFlattening
   * @param equatorialRadius in km
   */
  public constructor(inverseFlattening: number, equatorialRadius: number) {
    const CONST_RADIUS = 6371.0;

    this.equatorialRadius = equatorialRadius;
    this.inverseFlattening = inverseFlattening;

    if (Math.abs(inverseFlattening) === Infinity && inverseFlattening > 0) {
      this.flattening = 0;
      this.eccentricitySqr = 0;
      this.e1 = 1;
      this.e2 = 0;
      this.constantRadius = true;
    } else {
      this.flattening = 1 / inverseFlattening;
      this.eccentricitySqr = this.flattening * (2 - this.flattening);
      this.e1 = 1 - this.eccentricitySqr;
      this.e2 = this.eccentricitySqr / (1 - this.eccentricitySqr);
      this.constantRadius = this.equatorialRadius < CONST_RADIUS + 1.0;
    }
  }

  /**
   * Convert geographicLat in radians to geocentricLat in radians.
   *
   * @param geographicLat
   * @return geocentricLat in radians.
   */
  public getGeocentricLat(geographicLat: number): number {
    if (Math.abs(this.inverseFlattening) === Infinity) {
      return geographicLat;
    }

    return Math.atan(Math.tan(geographicLat) * this.e1);
  }

  /**
   * Get a unit vector corresponding to a point on the Earth with the specified latitude and
   * longitude.
   *
   * @param lat the geographic latitude, in radians.
   * @param lon the geographic longitude, in radians.
   */
  public getVector(lat: number, lon: number): [number, number, number] {
    const geocentricLat = this.getGeocentricLat(lat);
    return [
      Math.cos(geocentricLat) * Math.cos(lon),
      Math.cos(geocentricLat) * Math.sin(lon),
      Math.sin(geocentricLat)
    ];
  }

  /**
   * Get a unit vector corresponding to a point on the Earth with the specified latitude and
   * longitude.
   *
   * @param lat the geographic latitude, in degrees.
   * @param lon the geographic longitude, in degrees.
   * @return The returned unit vector.
   */
  public getVectorDegrees(lat: number, lon: number): number[] {
    return this.getVector(toRadians(lat), toRadians(lon));
  }
} // end of definition of enum EarthShape

export const EarthShapes = {
  /** The Earth is assumed to be a sphere of radius 6371 km. */
  SPHERE: new EarthShape(Infinity, EARTH_SPHERE_RADIUS),

  /**
   * The Earth is assumed to be an ellipsoid whose shape is defined by the GRS80 ellipsoid
   * specification.
   */
  GRS80: new EarthShape(GRS80_INVERSE_FLATTENING, EARTH_ELLIPSOID_RADIUS),

  /**
   * A hybrid coordinate system where latitudes are converted between geodetic and geocentric values
   * using the GRS80 ellipsoid, but conversions between depth and radius assume that the Earth has
   * constant radius of 6371 km.
   */
  GRS80_RCONST: new EarthShape(GRS80_INVERSE_FLATTENING, EARTH_SPHERE_RADIUS),

  /**
   * The Earth is assumed to be an ellipsoid whose shape is defined by the WGS84 ellipsoid
   * specification.
   */
  WGS84: new EarthShape(WGS84_INVERSE_FLATTENING, EARTH_ELLIPSOID_RADIUS),

  /**
   * A hybrid coordinate system where latitudes are converted between geodetic and geocentric values
   * using the WGS84 ellipsoid, but conversions between depth and radius assume that the Earth has
   * constant radius of 6371 km.
   */
  WGS84_RCONST: new EarthShape(WGS84_INVERSE_FLATTENING, EARTH_SPHERE_RADIUS),

  /**
   * The Earth is assumed to be an ellipsoid whose shape is defined by the IERS ellipsoid
   * specification.
   */
  IERS2003: new EarthShape(IERS_INVERSE_FLATTENING_FIVE, EARTH_ELLIPSOID_IERS_RADIUS),

  /**
   * A hybrid coordinate system where latitudes are converted between geodetic and geocentric values
   * using the IERS ellipsoid, but conversions between depth and radius assume that the Earth has
   * constant radius of 6371 km.
   */
  IERS2003_RCONST: new EarthShape(IERS_INVERSE_FLATTENING_FIVE, EARTH_SPHERE_RADIUS)
};
