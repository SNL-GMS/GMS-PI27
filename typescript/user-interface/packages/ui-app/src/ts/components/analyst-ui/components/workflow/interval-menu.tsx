import { Menu, MenuItem } from '@blueprintjs/core';
import type { WorkflowTypes } from '@gms/common-model';
import {
  isActivityInterval,
  isInteractiveAnalysisStageInterval,
  isStageInterval
} from '@gms/common-model/lib/workflow/types';
import type { ImperativeContextMenuProps } from '@gms/ui-core-components';
import { showImperativeReduxContextMenu, useAppSelector } from '@gms/ui-state';
import flatMap from 'lodash/flatMap';
import React from 'react';

import { getWorkflowPanelElement } from './workflow-util';

export interface IntervalMenuProps {
  readonly interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval;
  readonly isSelectedInterval: boolean;
  readonly allActivitiesOpenForSelectedInterval: boolean;

  readonly openCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  readonly closeCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
}

/**
 * Component that renders the interval context menu.
 */
export function IntervalMenu(props: IntervalMenuProps) {
  const openIntervalName = useAppSelector(state => state.app.workflow.openIntervalName);
  const {
    interval,
    isSelectedInterval,
    allActivitiesOpenForSelectedInterval,
    openCallback,
    closeCallback
  } = props;

  const userName = useAppSelector(state => state.app.userSession.authenticationStatus.userName);

  const isDisabled =
    (allActivitiesOpenForSelectedInterval &&
      isSelectedInterval &&
      openIntervalName === interval.intervalId.definitionId.name) ||
    (isSelectedInterval && interval.intervalId.definitionId.name !== openIntervalName);

  // determine if the logged in user is an active analyst; if so allow them to close that interval
  let isActiveAnalyst = false;
  if (isStageInterval(interval)) {
    if (isInteractiveAnalysisStageInterval(interval)) {
      isActiveAnalyst = flatMap(interval.activityIntervals.map(a => a.activeAnalysts)).includes(
        userName
      );
    }
  } else if (isActivityInterval(interval)) {
    isActiveAnalyst = interval.activeAnalysts.includes(userName);
  }

  return (
    <Menu>
      <MenuItem
        className="menu-item-open-interval"
        data-cy="open-interval-btn"
        text={isActiveAnalyst && !isDisabled ? 'Reopen interval' : 'Open interval'}
        disabled={isDisabled}
        onClick={() => openCallback(interval)}
      />
      <MenuItem
        className="menu-item-close-interval"
        data-cy="close-interval-btn"
        text="Close interval"
        disabled={!isSelectedInterval && !isActiveAnalyst}
        onClick={() => closeCallback(interval)}
      />
    </Menu>
  );
}

/**
 * Shows the {@link IntervalMenu} context Menu.
 *
 * @param event the event that invoked the action
 * @param props the {@link IntervalMenuProps} props
 * @param options (optional) imperative context menu options
 */
export const showIntervalMenu = (
  event: React.MouseEvent | MouseEvent,
  props: IntervalMenuProps,
  options: Pick<ImperativeContextMenuProps, 'activeElementOnClose' | 'onClose'> = {
    activeElementOnClose: getWorkflowPanelElement(),
    onClose: undefined
  }
) => {
  const {
    allActivitiesOpenForSelectedInterval,
    closeCallback,
    interval,
    isSelectedInterval,
    openCallback
  } = props;
  event.preventDefault();
  showImperativeReduxContextMenu({
    content: (
      <IntervalMenu
        allActivitiesOpenForSelectedInterval={allActivitiesOpenForSelectedInterval}
        closeCallback={closeCallback}
        interval={interval}
        isSelectedInterval={isSelectedInterval}
        openCallback={openCallback}
      />
    ),
    ...options,
    targetOffset: {
      left: event.clientX,
      top: event.clientY
    }
  });
};
