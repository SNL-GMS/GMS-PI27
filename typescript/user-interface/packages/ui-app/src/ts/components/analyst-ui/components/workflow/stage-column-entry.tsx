import type { WorkflowTypes } from '@gms/common-model';
import React from 'react';

import { SequenceIntervalCell } from './sequence-interval-cell';

export interface StageColumnEntryProps {
  readonly stageInterval: WorkflowTypes.StageInterval;
  readonly workflow: WorkflowTypes.Workflow | undefined;
}

function StageColumnEntryComponent(props: React.PropsWithChildren<StageColumnEntryProps>) {
  const { stageInterval, workflow } = props;
  return (
    <SequenceIntervalCell
      stageInterval={stageInterval}
      key={stageInterval.intervalId.definitionId.name}
      workflow={workflow}
    />
  );
}

/**
 * Row entry wrapper for Sequence cells
 */
export const StageColumnEntry: React.FunctionComponent<StageColumnEntryProps> =
  React.memo(StageColumnEntryComponent);
