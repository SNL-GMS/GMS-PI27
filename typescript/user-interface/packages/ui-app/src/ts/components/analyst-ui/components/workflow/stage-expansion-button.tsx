import { Button } from '@blueprintjs/core';
import React from 'react';

export interface StageExpansionButtonProps {
  readonly isExpanded: boolean;
  readonly disabled: boolean;
  readonly stageName: string;
  readonly toggle: () => void;
}

function StageExpansionButtonComponent(props: StageExpansionButtonProps) {
  const { disabled, stageName, isExpanded, toggle } = props;
  const smallSign = isExpanded ? 'small-minus' : 'small-plus';
  return !disabled ? (
    <Button
      key={stageName}
      className="stage-row__expand-button"
      icon={smallSign}
      onClick={toggle}
      disabled={disabled}
    />
  ) : null;
}

export const StageExpansionButton: React.FunctionComponent<StageExpansionButtonProps> = React.memo(
  StageExpansionButtonComponent
);
