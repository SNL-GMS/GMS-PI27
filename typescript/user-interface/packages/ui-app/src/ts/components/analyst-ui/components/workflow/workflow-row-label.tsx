import classNames from 'classnames';
import React from 'react';

export interface WorkflowRowLabelProps {
  readonly label: string;
  readonly isActivityRow: boolean;
}

export function WorkflowRowLabel(props: WorkflowRowLabelProps) {
  const { label, isActivityRow } = props;
  return (
    <div
      key={label}
      data-cy={`workflow-table-label-${label}`}
      className={classNames('workflow-table-label', {
        'workflow-table-label--activity': isActivityRow
      })}
    >
      <div>{label}</div>
    </div>
  );
}
