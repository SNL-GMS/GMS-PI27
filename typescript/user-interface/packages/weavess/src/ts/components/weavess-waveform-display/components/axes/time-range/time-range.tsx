import { WeavessConstants } from '@gms/weavess-core';
import * as React from 'react';

import { deltaTimeString, timeRangeDisplayString } from '../../../utils';
import type { TimeRangeProps } from './types';

function InternalTimeRange(props: TimeRangeProps) {
  const { labelWidthPx, currentZoomInterval } = props;
  const fullTimeString = timeRangeDisplayString(currentZoomInterval);
  const shortTimeString = timeRangeDisplayString(currentZoomInterval, 'HH:mm:ss.SSS');
  const veryShortTimeString = deltaTimeString(currentZoomInterval);
  return (
    <div
      className="weavess-wp-time-range"
      style={{
        left: `${labelWidthPx ?? WeavessConstants.DEFAULT_LABEL_WIDTH_PIXELS}px`
      }}
    >
      <span className="very-short-time">{veryShortTimeString}</span>
      <span className="short-time">{shortTimeString}</span>
      <span className="full-time">{fullTimeString}</span>
    </div>
  );
}

/**
 * A memoized component that simply draws a time range string at the desired location on the page.
 * If labelWidthPx are given, will draw them at that offset. Otherwise, defaults to WeavessConstants.DEFAULT_LABEL_WIDTH_PIXELS
 */
export const TimeRange = React.memo(InternalTimeRange);
