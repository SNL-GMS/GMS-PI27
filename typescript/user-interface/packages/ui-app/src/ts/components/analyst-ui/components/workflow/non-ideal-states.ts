import { Endpoints } from '@gms/common-model';
import { capitalizeFirstLetters } from '@gms/common-util';
import type { NonIdealStateDefinition } from '@gms/ui-core-components';
import {
  nonIdealStateWithError,
  nonIdealStateWithNoSpinner,
  nonIdealStateWithSpinner
} from '@gms/ui-core-components';
import type { StageIntervalsByIdAndTimeQuery, WorkflowQuery } from '@gms/ui-state';

export const workflowQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Data')
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return (
        props.workflowQuery?.isError &&
        (props.workflowQuery?.error?.message?.includes('404') ||
          props.workflowQuery?.error?.message?.includes('503'))
      );
    },
    element: nonIdealStateWithError(
      'Error',
      `Unable to communicate with ${capitalizeFirstLetters(
        Endpoints.WorkflowManagerServiceUrls.workflow.friendlyName
      )} Service`
    )
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isError;
    },
    element: nonIdealStateWithError('Error', 'Problem loading workflow data')
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return (
        !props.workflowQuery?.isLoading &&
        !props.workflowQuery?.isError &&
        !props.workflowQuery?.data &&
        props.workflowQuery?.data.stages.length === 0
      );
    },
    element: nonIdealStateWithNoSpinner('No Data', 'Workflow stages returned empty', 'exclude-row')
  }
];

export const workflowIntervalQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: {
      readonly hasFetchedInitialIntervals: boolean;
      workflowIntervalQuery: StageIntervalsByIdAndTimeQuery;
    }): boolean => {
      // only show the loading state if the on the initial fetch of the interval data
      return !props.hasFetchedInitialIntervals && props.workflowIntervalQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Interval Data')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return (
        !props.workflowIntervalQuery?.isLoading &&
        !props.workflowIntervalQuery?.isError &&
        (!props.workflowIntervalQuery.data || props.workflowIntervalQuery.data.length === 0)
      );
    },
    element: nonIdealStateWithNoSpinner('No Data', 'Stage intervals returned empty', 'exclude-row')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return props.workflowIntervalQuery?.isError;
    },
    element: nonIdealStateWithError('Error', 'Problem loading workflow interval data')
  }
];
