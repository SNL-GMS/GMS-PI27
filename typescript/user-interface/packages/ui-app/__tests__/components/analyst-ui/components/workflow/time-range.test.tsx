/* eslint-disable @typescript-eslint/no-magic-numbers */
import { SECONDS_IN_HOUR } from '@gms/common-util';
import { render } from '@testing-library/react';
import React from 'react';

import type { TimeRangeProps } from '../../../../../src/ts/components/analyst-ui/components/workflow/time-range';
import { TimeRange } from '../../../../../src/ts/components/analyst-ui/components/workflow/time-range';

// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

const secondsForDayPlus10 = SECONDS_IN_HOUR * 24 + 10;

const props: TimeRangeProps = {
  startTime: 0,
  endTime: secondsForDayPlus10
};

describe('workflow time axis tests', () => {
  it('is exported', () => {
    expect(TimeRange).toBeDefined();
  });

  it('matches snapshot', () => {
    const component = render(<TimeRange startTime={props.startTime} endTime={props.endTime} />);
    expect(component.baseElement).toMatchSnapshot();
  });
});
