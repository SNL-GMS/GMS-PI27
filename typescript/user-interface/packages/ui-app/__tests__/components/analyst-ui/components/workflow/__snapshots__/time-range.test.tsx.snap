// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow time axis tests matches snapshot 1`] = `
<body>
  <div>
    <div
      class="workflow-time-range"
    >
      <span
        class="time-range-left"
      >
        1970-01-01
      </span>
      <span
        class="time-range-right"
      >
        1970-01-02
      </span>
    </div>
  </div>
</body>
`;
